### PR TITLE
Irdl dialect

### PR DESCRIPTION
**Pull request by purvi-H**
Wrote the IRDL dialect in Python using the list of operations and attributes defined in https://github.com/opencompl/dyn-dialect + wrote cmath as an IRDL program.
Essentially, the important changes are in these files:
1. xdsl/src/xdsl/dialects/irdl.py where the irdl dialect is written
2. xdsl/src/xdsl/xdsl_opt_main.py where irdl dialect was imported and registered 
3. xdsl/tests/filecheck/cmath.xdsl where cmath written as IRDL program

### CODE REVIEW

File: src/xdsl/dialects/irdl.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        self.ctx.register_op(OperationOp)
+
+
+@irdl_attr_definition
+class EqTypeConstraintAttr(ParametrizedAttribute):
+
+    name = "equality_type_constraint"
+    data: ParameterDef[Attribute]
+
+    def enforce_one_param(data: Attribute) -> EqTypeConstraintAttr:
```
Comment (webmiche) [inline src/xdsl/dialects/irdl.py:38]: I don't quite see why this is needed or what this is used for. Shouldn't this be a verifier?

Also notice that you are not returning an `EqTypeCostraintAttr` but rather an `Attribute`.

File: src/xdsl/dialects/irdl.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        self.ctx.register_op(ConstraintVarsOp)
+        self.ctx.register_op(OperandsOp)
+        self.ctx.register_op(ResultsOp)
+        self.ctx.register_op(OperationOp)
+
+
+@irdl_attr_definition
+class EqTypeConstraintAttr(ParametrizedAttribute):
+
+    name = "equality_type_constraint"
```
[the remark this answers is not in the document]
Comment (purvi-H) [reply src/xdsl/dialects/irdl.py:35]: Yeah, just realised that the attributes are given the same name but the operations are prefixed with the dialect. I'll make this quick change. 

Will this also apply to the operation in builtin? i see that it's missing the prefix ```builtin.```

File: src/xdsl/dialects/builtin.py [hunk 3]
```diff
@@ -382,6 +382,13 @@ class Float32Type(ParametrizedAttribute):
 f32 = Float32Type()
 
 
+class Float64Type(ParametrizedAttribute):
```
[the remark this answers is not in the document]
Comment (purvi-H) [reply src/xdsl/dialects/builtin.py:385]: Oops, by mistake some changes relevant to my other PR trickled into this one :)) Consider this resolved.

File: src/xdsl/dialects/irdl.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (webmiche) [reply src/xdsl/dialects/irdl.py:35]: I know that MLIR is not entirely consistent here either (`llvm.struct` or `async.group` vs. `f32` or `i32`). IMO, the builtin types can be not prefaced because the builtin dialect is special either way considering that it provides things that everyone needs anyhow. But I don't have a strong opinion, maybe @math-fehr can indicate what he'd prefer?

File: src/xdsl/xdsl_opt_main.py [hunk 4]
```diff
@@ -76,7 +77,6 @@ def run(self):
     def register_all_arguments(self, arg_parser: argparse.ArgumentParser):
         """
         Registers all the command line arguments that are used by this tool.
-
```
Comment (webmiche) [inline src/xdsl/xdsl_opt_main.py:79]: There are still missing linebreaks in this file :)

Comment (purvi-H) [reply src/xdsl/xdsl_opt_main.py]: I've tried to fix them manually since I was having issues w yapf. Feel free to alert me if its missing in other places.

File: src/xdsl/xdsl_opt_main.py [hunk 5]
```diff
@@ -201,7 +202,7 @@ def _output_mlir(prog: ModuleOp, output: IOBase):
     def setup_pipeline(self):
         """
         Creates a pipeline that consists of all the passes specified.
-
```
Comment (math-fehr) [inline src/xdsl/xdsl_opt_main.py:204]: I guess this change is unrelated? Could you remove this?

File: src/xdsl/xdsl_opt_main.py [hunk 6]
```diff
@@ -137,7 +138,6 @@ def register_all_arguments(self, arg_parser: argparse.ArgumentParser):
     def register_all_dialects(self):
         """
         Register all dialects that can be used.
-
```
Comment (math-fehr) [inline src/xdsl/xdsl_opt_main.py:140]: I guess this change is unrelated? Could you remove this?

**Review by math-fehr — COMMENTED** [review]
Nice! I put some comments on it, but it's mostly renaming some fields.

File: src/xdsl/xdsl_opt_main.py [hunk 7]
```diff
@@ -137,7 +138,7 @@ def register_all_arguments(self, arg_parser: argparse.ArgumentParser):
...
+
```
Comment (math-fehr) [inline src/xdsl/xdsl_opt_main.py:141]: Could you remove the trailing spaces?

**Review by math-fehr — APPROVED** [review]
Thanks a lot, I just added two nit comments, then that should be good!
Could you also rebase from main?

File: src/xdsl/xdsl_opt_main.py (hunk 7 above)
Comment (math-fehr) [reply src/xdsl/xdsl_opt_main.py:141]: I think you might have resolved it by mistake, it's still here ^^
If you want I can remove it then merge the PR, but I don't want to complexify your git rebase!

Comment (purvi-H) [reply src/xdsl/xdsl_opt_main.py:141]: Oh yes, thanks for catching that. I attempted a rebase but one of the test fails now. We can look at it together tom and i can resolve this as well then.

**Comment by tobiasgrosser** [comment]
Nice! Congratulations @purvi-H for this big patch.
